return bad function http resp error

rdallman opened the pull request:
this was being thrown into the fn server logs but it's relatively easy to get
this to crop up if a function user forgets that they left a `println` laying
around that gets written to stdout, it garbles the http (or json, in its case)
output and they just see 'internal server error'. for certain clients i could
see that we really do want to keep this as 'internal server error' but for
things like e.g. docker image not authorized we're showing that in the
response, so this seems apt.

json likely needs the same treatment, will file a bug.

as always, my error messages are rarely helpful enough, help me please :)

closes #355

rdallman commented:
https://github.com/fnproject/fn/issues/729 for jason, it's a good first issue for somebody capable of growing sufficient beard to want to dig into the internals (really, I do not have the beard for this one...)

skinowski commented on api/agent/protocol/http.go:
```diff
@@ -33,6 +37,10 @@ func (h *HTTPProtocol) Dispatch(ctx context.Context, ci CallInfo, w io.Writer) e
 
 	resp, err := http.ReadResponse(bufio.NewReader(h.out), ci.Request()) // TODO timeout
 	if err != nil {
+		// From here (boo stdlib): https://github.com/golang/go/blob/d3c1df712658398f29bc8bebd6767e7b3cac2d12/src/net/http/response.go#L174
```
I don't think we should cherry pick based on error string match. There are many error variations in the stdlib (3-4 cases of malformed http only) so I think we should do one of the following:

1) mask all errors coming from ReadResponse as "invalid response from function" (hides internals, but better than generic 500)
2) pass through all errors from  ReadResponse inside NewAPIError (more user friendly, but could expose internal details)

Should these errors be 502 also?

IMHO, I'd pick '1' but using 502 bad gateway. (Since we are using 504 for timeout already and acting as a gw.)

rdallman replied:
at least one exception is that if the pipe closes they'll get io.ErrClosedPipe, but in theory that error won't get plumbed up since it's past the timeout (at some point, it may on accident leak out). 

i was hoping for a typed error from stdlib, sadly it's not the case. 

502 makes sense, 1) makes sense. we can log the real error. the error we get back conveniently does have some info in it such as `invalid HTTP status code: "MYOOPSYLOGLINE"` which could tip a user off as to what they're doing wrong. 

I was thinking it's possible since we're basically only doing i/o that we could plumb up any error from Dispatch as a `models.APIError` but the surface area may be too large. it would cover the json case nicely and also cover `req.Write` failures, but it's less clear what the exact error should be. something along the lines of "container i/o failed".

skinowski replied:
"invalid response from function", "cannot process function response" or something like that maybe? "container i/o failed" sounds good too except maybe that sounds like it could also be fn server issue.

rdallman replied:
I had another thought about this the other day that in http / json we could probably handle cases where uses write accidental log lines to stdout before the actual http / json blob. i'm not sure how worth it that is, but it's definitely a lego in the foot at the moment. it may well be better to try to get them to correct their behavior than covering up their misbehaving. 'invalid function response' kind of only makes sense in the case handled in this patch.

denismakogon replied:
agree with pt.1 - if function writes bad http response (whatever except well-structured http response) Fn should be more verbose for this cases and i'd like to get "invalid response from function" because it also indicates broken FDK as well as a user's faults.

rdallman replied:
it's more a matter of where to put it at this point. if we catch every error and say that it's not necessarily true. we're running on a hijacked connection, any number of things could go wrong there that aren't the functions fault, hence checking for very specific errors.

skinowski commented on api/agent/protocol/http.go:
```diff
@@ -33,6 +37,10 @@ func (h *HTTPProtocol) Dispatch(ctx context.Context, ci CallInfo, w io.Writer) e
 
 	resp, err := http.ReadResponse(bufio.NewReader(h.out), ci.Request()) // TODO timeout
```
req.Write() failure perhaps should be a 502.

skinowski commented:
Also, I'm seeing 502 from fn server for responses like this. Somewhere we must be replacing 500 with 502.

denismakogon commented:
The only thing i miss here is the test that does the following: creates a function with json or http format but the function is capable to do only default protocol framing (dummy echo func would be more than enough here). Overall, looks good.

rdallman commented:
yea, could use the hello image, can't use the test-utils image without being able to inject bad behavior into it. been marinating on exactly what scope this should cover in light of discussion, open to thoughts

skinowski commented:
I'm working on fn-test-utils to do stuff like this.. PR coming soon.

skinowski commented:
For now, you can trigger malformed http via fn-test-utils request:

```
{
        "responseCode": -1,
        "echoContent": "hello",
        "isDebug": true
}
```

skinowski commented:
I've got a #755 PR that has an invalid http case for this change.

rdallman commented on api/server/runner_test.go:
```diff
@@ -164,21 +166,17 @@ func TestRouteRunnerExecution(t *testing.T) {
-		{"/r/myapp/", ok, "GET", http.StatusOK, expHeaders, ""},
-		{"/r/myapp/myroute", ok, "GET", http.StatusOK, expHeaders, ""},
-		{"/r/myapp/myerror", crasher, "GET", http.StatusBadGateway, expHeaders, "container exit code 2"},
-		{"/r/myapp/myerror", crasher, "GET", http.StatusBadGateway, expHeaders, "container exit code 2"},
```
these were all duplicates except oom, it was time for these to go..

rdallman commented:
ok, i fixed up test and went ahead and fixed #729, thanks @skinowski for test image updates. should be ready to go... it does leak out the underlying error, in the prominent cases users keep hitting I do think it will help with specificity, even though extraneous cases are possible they still seem somewhat useful so that users can see where it went wrong.

rdallman commented:
thanks!
